### PR TITLE
feat: Deleting the common PVC if there are no workspaces left for a given user

### PR DESCRIPTION
**Pull request by dkwon17**
### What does this PR do?
Deletes the common PVC if there are no workspaces left for a given user. This PR copies over the three commits from this PR: https://github.com/eclipse/che/pull/18713.

### Screenshot/screencast of this PR
che-18369
After the three workspaces are deleted, the common PVC is deleted.

**PVCs are also deleted when `pvcStrategy` is set to `per-workspace`:**
che-18369_2
If `pvcStrategy` is set to `per-workspace`, after a workspace is deleted, the corresponding PVC is deleted.

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/18369

### How to test this PR?

The image: `quay.io/dkwon17/che-server:che-18369`

**Testing with helm installer:**
```
minikube start --addons=ingress --vm=true --memory=8192
chectl server:deploy --installer helm -p minikube --cheimage=quay.io/dkwon17/che-server:che-18369
```

**Testing with operator installer:**
When Che is deployed with the operator, we need this PR to provide delete permissions for the Che service account: https://github.com/eclipse-che/che-operator/pull/847.
My `che-operator` image for that PR is: `quay.io/dkwon17/che-operator:19867`.
```
minikube start --addons=ingress --vm=true --memory=8192
chectl server:deploy --installer operator -p minikube --cheimage=quay.io/dkwon17/che-server:che-18369 --che-operator-image=quay.io/dkwon17/che-operator:19867
```

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Comment by dkwon17** [comment]
@amisevsk @ibuziuk I had trouble reproducing this possible corner case: https://github.com/eclipse/che/pull/18713#pullrequestreview-561438604.

For the common strategy, I delete the Go workspace, and immediately create a Java Maven workspace, I don't seem to experience any blocking due to PVC deletion: 
che-18369_4

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java [hunk 1]
```diff
@@ -258,4 +274,70 @@ private PersistentVolumeClaim replacePVCsWithCommon(
...
+
+  /**
+   * The method is expected to be used in order to identify if the common PVC is used across all the
+   * users. The common PVC for all the users will be used if
+   * 'che.infra.kubernetes.namespace.default' points to a particular namespace e.g. 'che',
+   * 'workspaces' etc.
+   *
+   * @return true, if 'che.infra.kubernetes.namespace.default' is defined and does NOT contain the
+   *     placeholder e.g.: che-workspace-<username>), false otherwise
+   */
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:302]: is this code still valid? Since now we have only per-user namespace strategy and <username> or <userid> placeholder is mandatory.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java [hunk 2]
```diff
@@ -258,4 +274,70 @@ private PersistentVolumeClaim replacePVCsWithCommon(
...
+    if (!subject.isAnonymous()) {
+      String userId = subject.getUserId();
+      try {
+        Page<WorkspaceImpl> workspaces = workspaceManager.getWorkspaces(userId, false, 1, 0);
+        if (workspaces.isEmpty()) {
+          log.debug("User '{}' has no more workspaces left", userId);
+          return true;
+        }
+      } catch (ServerException e) {
+        log.error(
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:334]: This is an impossible sutuation. I believe we should throw a RuntimeException here. Or remove this block at all.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java [hunk 3]
```diff
@@ -258,4 +274,70 @@ private PersistentVolumeClaim replacePVCsWithCommon(
...
+        && !defaultNamespaceName.contains("<")
+        && !defaultNamespaceName.contains(">");
+  }
+
+  /** @return true, if there are no workspaces left across all the users, false otherwise */
+  private boolean totalNumberOfWorkspacesIsZero() {
+    try {
+      return workspaceManager.getWorkspacesTotalCount() == 0;
+    } catch (ServerException e) {
+      log.error("Unable to get the total number of workspaces. Cause: {}", e.getMessage(), e);
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:314]: why this error has been suppressed?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PerWorkspacePVCStrategy.java [hunk 4]
```diff
@@ -57,25 +59,29 @@ public PerWorkspacePVCStrategy(
       @Named("che.infra.kubernetes.pvc.precreate_subpaths") boolean preCreateDirs,
       @Named("che.infra.kubernetes.pvc.storage_class_name") String pvcStorageClassName,
       @Named("che.infra.kubernetes.pvc.wait_bound") boolean waitBound,
+      @Nullable @Named("che.infra.kubernetes.namespace.default") String defaultNamespaceName,
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PerWorkspacePVCStrategy.java:62]: it's mandatory property afaik

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java [hunk 5]
```diff
@@ -279,8 +294,15 @@ public void shouldDeletePVCsIfThereIsNoPersistAttributeInWorkspaceConfigWhenClea
       throws Exception {
     // given
     Workspace workspace = mock(Workspace.class);
+    Page workspaces = mock(Page.class);
+
     lenient().when(workspace.getId()).thenReturn(WORKSPACE_ID);
 
+    lenient()
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java:301]: why do you add `lenient()`  inside the test?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java (hunk 1 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:302]: @skabashnyuk nice catch, it looks like we do not need this check anymore - https://github.com/eclipse/che/pull/19409/files

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java (hunk 3 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:314]: technically, this error should never happen. If it happens we just log error and return false (do not delete PVC)

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java (hunk 5 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java:301]: could you clarify a bit, seems to be smth that is used all over the shop

**Review by ibuziuk — COMMENTED** [review]
@dkwon17 could you please follow the brand new semantic PR requirment https://github.com/zeke/semantic-pull-requests

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java (hunk 1 above)
Comment (dkwon17) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java]: Thank you for the feedback, I removed the method and the two tests for this check

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PerWorkspacePVCStrategy.java (hunk 4 above)
Comment (dkwon17) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PerWorkspacePVCStrategy.java]: Removed `@Nullable`

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java (hunk 2 above)
Comment (dkwon17) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java]: I now throw a RuntimeException here

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java (hunk 3 above)
Comment (dkwon17) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java]: Should I throw a RuntimeException here to be consistent with https://github.com/eclipse-che/che-server/pull/16#discussion_r641895327?  Or maybe keep `log.error()` for both cases?

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java (hunk 5 above)
Comment (skabashnyuk) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java:300]: lenient is a workaround. That sometimes used in setup methods. It means that this stubbing might be unused in some tests. I see no reason to add useless stubbing inside the test. Or maybe it not useless?  In that case, I do not need lenient.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java [hunk 6]
```diff
@@ -258,4 +272,31 @@ private PersistentVolumeClaim replacePVCsWithCommon(
...
+  }
+
+  /**
+   * @return true, if a given user has no workspaces, false otherwise (or if the subject is
+   *     anonymous)
+   * @throws RuntimeException
+   */
+  private boolean userHasNoWorkspaces() {
+    Subject subject = EnvironmentContext.getCurrent().getSubject();
+    if (!subject.isAnonymous()) {
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:287]: you don't need this if.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java [hunk 7]
```diff
@@ -258,4 +272,31 @@ private PersistentVolumeClaim replacePVCsWithCommon(
...
+      String userId = subject.getUserId();
+      try {
+        Page<WorkspaceImpl> workspaces = workspaceManager.getWorkspaces(userId, false, 1, 0);
+        if (workspaces.isEmpty()) {
+          log.debug("User '{}' has no more workspaces left", userId);
+          return true;
+        }
+      } catch (ServerException e) {
+        // should never happen
+        throw new RuntimeException(e.getLocalizedMessage(), e);
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:297]: Just wonder. Maybe it's better to wrap it in `InfrastructureException` wdyt?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java (hunk 6 above)
Comment (dkwon17) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java]: Thank you, it is now removed

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java [hunk 8]
```diff
@@ -422,44 +422,6 @@ public void shouldNotDeleteCommonPVCIfUserHasWorkspaces() throws Exception {
...
-
-    // when
-    commonPVCStrategy.cleanup(workspace);
-
-    // then
-    verify(workspaceManager, never()).getWorkspaces(anyString(), eq(false), anyInt(), anyLong());
-    verify(ns, never()).persistentVolumeClaims();
-    verify(persistentVolumeClaims, never()).delete(PVC_NAME);
-    verify(pvcSubPathHelper).removeDirsAsync(WORKSPACE_ID, "ns", PVC_NAME, WORKSPACE_ID);
-  }
```
Comment (dkwon17) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java]: In reference to the if check from https://github.com/eclipse-che/che-server/pull/16#discussion_r642919396, I deleted this test.

It fails without the if check, due to an NPE (workspace is `null`) within the code that was inside the if check's body here: https://github.com/eclipse-che/che-server/blob/f01607926ffbeac3a9978bfd5fcdbd88d5ff4c92/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java#L290

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java (hunk 7 above)
Comment (dkwon17) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java]: sounds good, fixed

**Comment by skabashnyuk** [comment]
[crw-ci-test --rebuild]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java [hunk 9]
```diff
@@ -258,4 +272,29 @@ private PersistentVolumeClaim replacePVCsWithCommon(
...
+    factory.get(workspace).persistentVolumeClaims().delete(configuredPVCName);
+  }
+
+  /**
+   * @return true, if a given user has no workspaces, false otherwise
+   *
+   * @throws InfrastructureException
+   */
+  private boolean userHasNoWorkspaces() throws InfrastructureException {
+    Subject subject = EnvironmentContext.getCurrent().getSubject();
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:286]: This might throw an NPE in a generic case. Because `cleanup` is called just during an event handler, it might be called in a situation where there is no initiating REST call that would set the subject.

Also, generically speaking, the workspace is owned by an account, not a user, so it might belong to an organization. While we no longer support that, I think, in the UI, we should take that possibility into account (pun not intended) here.

So instead of using `EnvironmentContext` to get the current user, I would instead get the owner of the workspace using the (appropriately named) `getNamespace()` method on the workspace, check if that is indeed a user, not an organization, and then proceed from there.

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java [hunk 10]
```diff
@@ -324,6 +348,127 @@ public void shouldDeletePVCsIfPersistAttributeIsSetToTrueInWorkspaceConfigWhenCl
...
+    commonPVCStrategy.cleanup(workspace);
+
+    // then
+    verify(ns, never()).persistentVolumeClaims();
+    verify(persistentVolumeClaims, never()).delete(PVC_NAME);
+    verify(pvcSubPathHelper).removeDirsAsync(WORKSPACE_ID, "ns", PVC_NAME, WORKSPACE_ID);
+  }
+
+  @Test
+  public void shouldNotCheckTotalNumberOfWorkspacesIfDefaultNamespaceContainsPlaceholder()
```
Comment (metlos) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java:426]: What is this testing? It doesn't seem related to the changes as far as I can tell...

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java (hunk 9 above)
Comment (skabashnyuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:286]: @metlos what about checking that ```EnvironmentContext.getCurrent().getSubject() == workspace.getAccount().getId()```

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java (hunk 10 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java:426]: should be removed since the check was related to https://github.com/eclipse-che/che-server/pull/16#discussion_r641895137

**Review by ibuziuk — APPROVED** [review]
LGTM, assuming that formatting is fixed (`mvn fmt:format` should be run according to the PR check) + the non-relevant test is removed https://github.com/eclipse-che/che-server/pull/16#discussion_r643767878

@dkwon17 please also provide a relevant PR to docs with a new feature description. Smth. like:

NOTE: PVC will be removed once all user's workspaces are deleted. PVC will be re-created once a new non-ephemeral workspace is started.

https://www.eclipse.org/che/docs/che-7/installation-guide/configuring-storage-strategies/

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java (hunk 10 above)
Comment (dkwon17) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategyTest.java]: thank you, fixed

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java (hunk 9 above)
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:285]: > @metlos what about checking that `EnvironmentContext.getCurrent().getSubject() == workspace.getAccount().getId()`

Well, this is an event handler, so I think it is safe to assume that whatever triggered the event has checked that whoever initiated the delete has enough privs. The user might not be in the env context at all (because theoretically this can get called from a codepath not originating in a REST request)... That's why I suggested to just look at the account of the workspace, and if it is a user, check if there is no more workspaces of that user and delete the pvc.

Comment (sparkoo) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:285]: It will not throw NPE, but return anonymous user. I'm not exactly sure what's the consequence of that here. https://github.com/eclipse-che/che-server/blob/main/core/che-core-api-core/src/main/java/org/eclipse/che/commons/env/EnvironmentContext.java#L56

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java [hunk 11]
```diff
@@ -105,12 +112,14 @@ public CommonPVCStrategy(
       @Named("che.infra.kubernetes.pvc.precreate_subpaths") boolean preCreateDirs,
       @Named("che.infra.kubernetes.pvc.storage_class_name") String pvcStorageClassName,
       @Named("che.infra.kubernetes.pvc.wait_bound") boolean waitBound,
+      @Nullable @Named("che.infra.kubernetes.namespace.default") String defaultNamespaceName,
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:115]: why is this needed? Looks like it's not used anywhere.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java (hunk 9 above)
Comment (skabashnyuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:285]: @metlos what about 
```java
@Override
  public void cleanup(Workspace workspace) throws InfrastructureException {
    if (EphemeralWorkspaceUtility.isEphemeral(workspace)) {
      return;
    } else if (userHasNoWorkspaces(((WorkspaceImpl) workspace).getAccount())) {
      log.debug("Deleting the common PVC: '{}',", configuredPVCName);
      deleteCommonPVC(workspace);
      return;
    }
    String workspaceId = workspace.getId();
    PersistentVolumeClaim pvc = createCommonPVC(workspaceId);
    pvcSubPathHelper.removeDirsAsync(
        workspaceId,
        factory.get(workspace).getName(),
        pvc.getMetadata().getName(),
        subpathPrefixes.getWorkspaceSubPath(workspaceId));
  }
....
  private boolean userHasNoWorkspaces(AccountImpl account) throws InfrastructureException {
    try {
      if("personal".equals(account.getType())) {
        Page<WorkspaceImpl> workspaces = workspaceManager.getWorkspaces(account.getId(), false, 1, 0);
        if (workspaces.isEmpty()) {
          log.debug("User '{}' has no more workspaces left", account.getId());
          return true;
        }
      }else{
        // should never happen in 7.x since organizations is not used.
        return false;
      }
    } catch (ServerException e) {
      // should never happen
      throw new InfrastructureException(e.getLocalizedMessage(), e);
    }
    return false;
  }
```

Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:285]: @skabashnyuk that is exactly what I had in mind. Thanks for stepping up and working out the details. 👍🏼

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java (hunk 11 above)
Comment (dkwon17) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java]: Oops, it was from a test that was removed. Fixed in the latest commit.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java (hunk 9 above)
Comment (dkwon17) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java]: Thanks everyone, I made some changes in the latest commit.

**Comment by skabashnyuk** [comment]
[crw-ci-test --rebuild]

**Review by sparkoo — APPROVED** [review]
great contribution, thanks

**Comment by skabashnyuk** [comment]
[crw-ci-test --rebuild]

**Review by skabashnyuk — APPROVED** [review]

**Review by metlos — APPROVED** [review]
Looks great, thank you!

**Comment by skabashnyuk** [comment]
[crw-ci-test --rebuild]

**Comment by dkwon17** [comment]
Thanks everyone. Is now a good time for me to squash all commits into one?

**Comment by nickboldt** [comment]
> Is now a good time for me to squash all commits into one?

Unless there's a compelling reason to show the commit history, I always use 'squash and merge' to apply PRs. 

Sometimes I'll manually rebase a bunch of commits from, say, 12 commits down to 2 or 3 (if the PR covers more than one interesting change) and then NOT squash-merge... but that's the exception. 

TL;DR, I'd say yes, squash away!
